### ISSUE TITLE
Composite primary key order is not consistent. Using the 'expo' driver in a React Native project

### ISSUE THREAD

**Issue by KFeu1** [drizzle-team/drizzle-kit-mirror#342]
### Description

I have two tables, *recipes* and *tags*, and I want to create another table *recipe_has_tags*, which should have a composite primary key, made from the primary keys of both tables.

I am on Windows 11, working on a React Native project, using the *expo* driver in my `drizzle.config.ts`.

Here is my `schema.ts`:

```ts
export const recipes = sqliteTable('recipes', {
	id: integer("id").primaryKey({ autoIncrement: true }),
  	name: text('name', { length: 256 }).notNull(),
});

export const tags = sqliteTable('tags', {
	id: integer("id").primaryKey({ autoIncrement: true }),
  	name: text('name', { length: 256 }).notNull(),
});

export const recipeHasTags = sqliteTable('recipe_has_tags', {
	recipeId: integer('recipe_id').references(() => recipes.id).notNull(),
	tagId: integer('tag_id').references(() => tags.id).notNull(),
}, (table) => {
	return {
		pk: primaryKey({ columns: [table.tagId, table.recipeId] }),
	}
});
```

### The problem

When I run `pnpm exec drizzle-kit generate:sqlite`, the resulting .sql file contains this line for the *recipe_has_tags* table:

```sql
PRIMARY KEY(`recipe_id`, `tag_id`)
```

but the `schema.ts` had those in reverse order. I want *tag_id* to be the first key of the composite key.

### Workaround

Renaming the columns, so that they are alphabetically in the correct order fixes this:

```ts
export const recipeHasTags = sqliteTable('recipe_has_tags', {
	recipeId: integer('pk2_recipe_id').references(() => recipes.id).notNull(),
	tagId: integer('pk1_tag_id').references(() => tags.id).notNull(),
}, (table) => {
	return {
		pk: primaryKey({ columns: [table.tagId, table.recipeId] }),
	}
});
```

This produces the result I want, which is

```sql
PRIMARY KEY(`pk1_tag_id`, `pk2_recipe_id`)
```

but obviously the order should not be derived from the names of the columns.

### How to reproduce

Here is my project in a zip file:

[migration_bug.zip](https://github.com/drizzle-team/drizzle-kit-mirror/files/14666730/migration_bug.zip)

To reproduce, just run `pnpm install` and then `pnpm migrate`. This creates the *drizzle* folder, where you can look at the migration it created.

**Comment by totto2727** [comment]
Not with Expo, but with D1, but I faced the same problem.
It should be generated in the same order as in the array.

```ts
export const rssTarget = sqliteTable(
  "rss_target",
  {
    userUuidv4: text("user_uuidv4")
      .references(() => users.uuidv4, referenceDefault)
      .notNull(),
    url: text("url").notNull(),
    name: text("name").notNull(),
    ...createdAtAndUpdatedAt,
  },
  (table) => ({
    userIdAndUrlPrimaryKey: primaryKey({
      columns: [table.userUuidv4, table.url],
    }),
  })
```

```sql
CREATE TABLE `rss_target` (
	`user_uuidv4` text NOT NULL,
	`url` text NOT NULL,
	`name` text NOT NULL,
	`created_at` text NOT NULL,
	`updated_at` text NOT NULL,
	PRIMARY KEY(`url`, `user_uuidv4`),
	FOREIGN KEY (`user_uuidv4`) REFERENCES `users`(`uuidv4`) ON UPDATE cascade ON DELETE no action
)
```

**Comment by chrisbrown-io** [comment]
After some digging through the `drizzle-kit` source, it seems this issue stems from the sqlite schema serialization that occurs when generating migrations (`drizzle-kit generate`), pushing schema changes (`drizzle-kit push`) and introspecting (`drizzle-kit introspect`).

When defining a composite primary key in your schema, the columns to include are specified as an array in your schema. It turns out that these are being sorted using the default `Array.sort` method (alphabetic ascending) during the conversion into the structure required for serialization. This is to ensure consistency between the output of generate/push commands and introspect.

I can't link to specific parts of the code, as `drizzle-kit` is closed source (😢 ), but this happens within the `generateSqliteSnapshot` method of `sqliteSerializer`:

```
primaryKeys.forEach((it) => {
          if (it.columns.length > 1) {
            primaryKeysObject[it.getName()] = {
              columns: it.columns.map((it2) => it2.name).sort(),
              name: it.getName()
            };
```

A fix for this is to remove the `.sort()`, and the resulting migration and push output will have composite primary key definitions in the specified array order within your schema 🎉  You can apply the following patch to `drizzle-kit@0.23.0`.

```patch
diff --git a/bin.cjs b/bin.cjs
index 5f1228dc9a361fcd0443146445b27c119a01bd01..ca6fb4090856bc8417979c859cc7c81285e5dcf0 100755
--- a/bin.cjs
+++ b/bin.cjs
@@ -22617,7 +22617,7 @@ The unique constraint ${source_default.underline.blue(
         primaryKeys.forEach((it) => {
           if (it.columns.length > 1) {
             primaryKeysObject[it.getName()] = {
-              columns: it.columns.map((it2) => it2.name).sort(),
+              columns: it.columns.map((it2) => it2.name),
               name: it.getName()
             };
           } else {
```

This only seems to affect sqlite(?) and I can't see the same sort logic being applied to PostgreSQL or MySQL.

I assume the explanation for including the sort, is due to the sqlite introspection serialization implementation. This currently queries all columns for all tables. The columns are iterated over and any column names that are marked as a primary key are added into an array for each table. The order that these columns are returned in and iterated over will not necessarily match the order of your composite primary key definition in your schema, so they are similarly being sorted:

```
      for (const [key, value] of Object.entries(tableToPk)) {
        if (value.length > 1) {
          value.sort();
          result[key].compositePrimaryKeys = {
            [`${key}_${value.join("_")}_pk`]: {
              columns: value,
              name: `${key}_${value.join("_")}_pk`
            }
          };
```

It's hard to properly diagnose without having access to the source, but I feel fix for the introspection serialization issue would be to query the indexes from the database (if this is possible?) and reconstruct the composite primary key columns that way . This would ensure the schema that is output has the correct composite primary key order. This falls outside my expertise however, so this may not be possible...

**Comment by AndriiSherman** [comment]
Should be fixed in `drizzle-kit@0.24.0`
Please check [release notes](https://github.com/drizzle-team/drizzle-orm/releases/tag/drizzle-kit%400.24.0) before updating
